tests/MongoDB.Driver.Core.Tests/Core/Connections/ConnectionInitializerTests.cs

```diff
@@ -67,6 +69,63 @@ public void InitializeConnection_should_acquire_connectionId_from_isMaster_respo
             result.ConnectionId.ServerValue.Should().Be(1);
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void InitializeConnection_should_speculatively_authenticate_with_SCRAM_SHA(
+            [Values("default", "SCRAM-SHA-256", "SCRAM-SHA-1")] string authenticatorType,
+            [Values(false, true)] bool async)
+        {
+            var isMasterReply = MessageHelper.BuildReply(
+                RawBsonDocumentHelper.FromJson("{ ok : 1, connectionId : 1 }"));
+            var buildInfoReply = MessageHelper.BuildReply(
+                RawBsonDocumentHelper.FromJson("{ ok : 1, version : \"4.2.0\" }"));
+            var credentials = new UsernamePasswordCredential(
+                source: "Voyager", username: "Seven of Nine", password: "Omega-Phi-9-3");
+            var authenticator = CreateAuthenticator();
```

Comment:
A switch expression here (https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#switch-expressions) would be nice, but we haven't enabled C#8 yet.